### ISSUE TITLE
[Python] Unable to instantiate oneOf when spec oneOf component schemas are not fully constrained

### ISSUE THREAD

**Issue by awildturtok** [OpenAPITools/openapi-generator#14255]
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

Writing a component that uses both `oneOf` and `discriminator` produces faulty code, duplicating the entries in the discriminator classmethod. This causes deserialization to fail as multiple sub-classes match any input.

Dropping the `oneOf` block produces the desired code and deserializes. However the generated TS code will now generate a `Set<any>` for array of `semantics` as semantics is erased(?).

##### openapi-generator version

6.3.0-SNAPSHOT via hub.docker.com/openapitools/openapi-generator-cli:latest

##### OpenAPI declaration file content or url

https://gist.github.com/awildturtok/31e9ccf175954b21e9f161ec40105a1f

##### Steps to reproduce

` java --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.util=ALL-UNNAMED -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g python -i openapi.yaml`

**Comment by spacether** [comment]
Thank you for reporting this. So your spec as-is was unable to generate a client.
I have added the missing pieces to create a spec that can be generated.

Your spec has an error where it defines `components.schemas.holder.properties.semantics.items.schema`
`schema` is not a valid openapi or json-schema schema keyword. I have replaced it with the deeper $ref.

The working spec that I used to generate a client is:
```

openapi: 3.0.3
info:
  title: some api
  version: 1.0.0
paths: {}
components:
  schemas:
    semantic-base:
      type: object
      properties:
        type:
          type: string
      required:
        - "type"
    semantic-hidden:
      allOf:
        - $ref: "#/components/schemas/semantic-base"
    semantic-id:
      allOf:
        - $ref: "#/components/schemas/semantic-base"
        - type: object
          properties:
            kind:
              type: string
    semantics:
      discriminator:
        propertyName: "type"
        mapping:
          "ID": "#/components/schemas/semantic-id"
          "HIDDEN": "#/components/schemas/semantic-hidden"
      oneOf:
        - $ref: "#/components/schemas/semantic-id"
        - $ref: "#/components/schemas/semantic-hidden"
    holder:
      properties:
        semantics:
          type: array
          uniqueItems: true
          items:
            $ref: "#/components/schemas/semantics"
```
And when I run this sample code:
```
from openapi_client.model import semantics

s = semantics.Semantics(
    type='ID'
)
```
I get this exception:
```
openapi_client.exceptions.ApiValueError: Invalid inputs given to generate an instance of <class 'openapi_client.model.semantics.Semantics'>. Multiple oneOf schemas [<class 'openapi_client.model.semantic_id.SemanticId'>, <class 'openapi_client.model.semantic_hidden.SemanticHidden'>] matched the inputs, but a max of one is allowed.
```
Which is a correct valid error. openapi/json schema keywords are validated separately.
Both SemanticId and SemanticHidden pass validation for this instance.
Only one of them must pass validation per the oneOf definition in `semantics`
To fix this, add a string enum value for type in both of those schemas like so:
```
openapi: 3.0.3
info:
  title: some api
  version: 1.0.0
paths: {}
components:
  schemas:
    semantic-base:
      type: object
      properties:
        type:
          type: string
      required:
        - "type"
    semantic-hidden:
      allOf:
        - $ref: "#/components/schemas/semantic-base"
      properties:
        type:
          type: string
          enum:
            - "HIDDEN"
    semantic-id:
      allOf:
        - $ref: "#/components/schemas/semantic-base"
        - type: object
          properties:
            kind:
              type: string
      properties:
        type:
          type: string
          enum:
            - "ID"
    semantics:
      discriminator:
        propertyName: "type"
        mapping:
          "ID": "#/components/schemas/semantic-id"
          "HIDDEN": "#/components/schemas/semantic-hidden"
      oneOf:
        - $ref: "#/components/schemas/semantic-id"
        - $ref: "#/components/schemas/semantic-hidden"
    holder:
      properties:
        semantics:
          type: array
          uniqueItems: true
          items:
            $ref: "#/components/schemas/semantics"
```
After re-generating the client with the above fixed spec, the code sample works and the oneOf class can be instantiated without errors.

**Comment by awildturtok** [comment]
> Your spec has an error where it defines components.schemas.holder.properties.semantics.items.schema
schema is not a valid openapi or json-schema schema keyword. I have replaced it with the deeper $ref.

Sorry that was a copy pasting error.
---
Your suggestion has worked, thank you!

However, if I'm understanding you right, this means that the discriminator is not actually used for code generation? I would have expected discriminator to imply limiting the scope of subtypes, or is that just down to how I composed the schema? If so I would be more than happy to restructure it.
